### ISSUE TITLE
feat: Barcode scanning for Web platform

### ISSUE THREAD

**Issue by piotr-cz** [capawesome-team/capacitor-mlkit#20]
**Is your feature request related to a problem? Please describe:**

N/A

**Describe the solution you'd like:**

As in description.

Documentation: mdn > [Barcode Detection API](https://developer.mozilla.org/en-US/docs/Web/API/Barcode_Detection_API)
Demo: <https://whatpwacando.today/barcode>

**Describe alternatives you've considered:**

N/A

**Additional context:**

N/A

**Comment by thegnuu** [comment]
@robingenz I can have a look at this if you want :) As you mentioned in our discussion you already have a standalone component relying on `BarcodeDetector`, unfortunately, support of `BarcodeDetector` is not optimal at the moment (https://caniuse.com/?search=BarcodeDetector).

My idea would be to use https://github.com/undecaf/barcode-detector-polyfill, I was already in contact with the guys over there to allow bundling of `zbar-wasm` which is possible now, and I think using a polyfill for not supported browsers would be a good idea at the current stage, what do you think about this? This might be a bit of an overhead but it should allow better and broader usage of this implementation and reduce "web platform usage" issues.

**Comment by robingenz** [comment]
@thegnuu That sounds like a great idea. What about the license? As far as I can see https://github.com/mchehab/zbar is licensed under `LGPL-2.1` and therefore not compatible with `Apache-2.0`.

**Comment by thegnuu** [comment]
@robingenz I am not entirely sure, to be honest. There was an issue over there https://github.com/undecaf/barcode-detector-polyfill/issues/2 and I think we should be able to use the plugin and bundle it according to this, but I am by far no license expert ;)

**Comment by robingenz** [comment]
@thegnuu Unfortunately, I am not an expert either. As far as I understand it, it gets complicated as soon as the LGPL source code is linked statically (see https://stackoverflow.com/a/10179181/6731412), which should be the case since the library is compiled into the app as part of the build process.

For example, another plugin author recently switched to the MPL license as there were concerns (siehe https://github.com/Cap-go/capacitor-updater/issues/7#issuecomment-1454893916).

**Comment by robingenz** [comment]
> For example, another plugin author recently switched to the MPL license as there were concerns (see Cap-go/capacitor-updater#7 (comment)).

@thegnuu I just read about the MPL license and see problems there as well:

> 3.2. Distribution of Executable Form
> If You distribute Covered Software in Executable Form then:
>
> such Covered Software must also be made available in Source Code Form, as described in Section 3.1, and You must > inform recipients of the Executable Form how they can obtain a copy of such Source Code Form by reasonable means in a timely manner, at a charge no more than the cost of distribution to the recipient; and
>
> You may distribute such Executable Form under the terms of this License, or sublicense it under different terms, provided that the license for the Executable Form does not attempt to limit or alter the recipients’ rights in the Source Code Form under this License.

[Source](https://www.mozilla.org/en-US/MPL/2.0/)

**Comment by thegnuu** [comment]
I see, I see. Another idea would be to just implement `BarcodeDetector` within this plugin and properly document it like this. If someone needs to use it in a not supported browser, they can implement the initialization of the polyfill on their end. Since there are multiple polyfills for `BarcodeDetector` (e.g. https://github.com/xulihang/barcode-detector-polyfill) this would allow each dev to choose the version they want.

It's a more generic approach of course and might not work out of the box on each browser, but it would get us around the "license struggles" and add more flexibility for the devs. Adding a polyfill depending on `zxing` like the example in this comment seems not to be the best idea IMHO.

**Comment by robingenz** [comment]
> I see, I see. Another idea would be to just implement BarcodeDetector within this plugin and properly document it like this. If someone needs to use it in a not supported browser, they can implement the initialization of the polyfill on their end. Since there are multiple polyfills for BarcodeDetector (e.g. [xulihang/barcode-detector-polyfill](https://github.com/xulihang/barcode-detector-polyfill)) this would allow each dev to choose the version they want.

Yes, i think this is the best solution atm. Any PR is welcome. Otherwise, I would put it on my todo list next month.

**Comment by piotr-cz** [comment]
My feature request was initially about adding just native *Barcode Detection API* for web platform.
Adding a note in docs about polyfills is definitely good idea.

**Comment by thegnuu** [comment]
I think this makes sense yes, unfortunately I have a really busy week, but maybe I will find some time to start the implementation.

**Comment by robingenz** [comment]
@thegnuu I had some time on the train today and started with the implementation, see #52.

**Comment by maelp** [comment]
Perhaps this code could be useful? It handles Datamatrix https://github.com/rxing-core/rxing-wasm/tree/main

**Comment by iBobik** [comment]
Hi, I’ve just published my PR for this: https://github.com/capawesome-team/capacitor-mlkit/pull/211

It also works well with the https://www.npmjs.com/package/barcode-detector polyfill which I choose for it’s easy usage event with offline capability.

**Comment by robingenz** [comment]
This feature is now available as pre-release:

```
npm i https://pkg.pr.new/capawesome-team/capacitor-mlkit/@capacitor-mlkit/barcode-scanning@238
```

Thanks to @iBobik for his work!

**Comment by iBobik** [comment]
@robingenz Glad to hear that.

It will take time until I will test your version, just one note about potential bug:

Previously there was similar approach with the video element - passing it from the options, but I was not able to make it working in native plugins because DOM object cannot be serialized. So it required pretty tricky and unintuitive configuration - something like:
```
videoEl: Capacitor.isNative() ? undefined : videoEl
``` 

This was the reason why I wrote web plugin to behave the same like native - create video element itself.

**Comment by robingenz** [comment]
@iBobik Thank you for your feedback. I will include this in the `Usage` section. I think there will always be differences between web and native. In the end, it's a personal preference which one you like better.
